# Q2

In [1]:
!python ../template/model.py

LeNet5(
  (conv1): Conv2d(1, 6, kernel_size=(5, 5), stride=(1, 1))
  (act1): ReLU()
  (pool1): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (conv2): Conv2d(6, 16, kernel_size=(5, 5), stride=(1, 1))
  (act2): ReLU()
  (pool2): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (fc1): Linear(in_features=256, out_features=120, bias=True)
  (act3): ReLU()
  (fc2): Linear(in_features=120, out_features=84, bias=True)
  (act4): ReLU()
  (fc3): Linear(in_features=84, out_features=10, bias=True)
)
([150, 6, 2400, 16, 30720, 120, 10080, 84, 840, 10], 44426)
----------------------
CustomMLP(
  (fc1): Linear(in_features=784, out_features=54, bias=True)
  (act1): ReLU()
  (fc2): Linear(in_features=54, out_features=28, bias=True)
  (act2): ReLU()
  (fc3): Linear(in_features=28, out_features=10, bias=True)
)
([42336, 54, 1512, 28, 280, 10], 44220)


## LeNet-5 Model Parameters
### Conv1
- Input Channel : 1
- Filter : 6
- Kernel : 5 * 5
- Num of Parameters = 1 * 6 * 5 * 5 = 150
- Bias(same with Activation) = 6
### Conv2
- Input Channel : 6 (Conv1 Output)
- Filter : 16
- Kernel : 5 * 5
- Num of Parameters = 16 * 6 * 5 * 5
- Bias = 16
### FC1
- Input : 16 * 4 * 4
- Output : 120
- Num of Parameters = 16 * 4 * 4 * 120 = 30720
- Bias = 120

### FC2
- Input : 120
- Output : 84
- Num of Parameters = 120 * 84 = 10080
- Bias = 84

### FC3
- Input : 84
- Output : 10
- Num of Parameters = 84 * 10 = 840
- Bias = 10

### Total : 150 + 6 + 2400 + 16 + 30720 + 120 + 10080 + 84 + 840 + 10 = 44426

## CustomMLP Model Parameters
### FC1
- Input : 28 * 28
- Output : 54
- Num of Parameters = 28 * 28 * 54 = 42336
- Bias = 54

### FC2
- Input : 54
- Output : 28
- Num of Parameters = 54 * 28 = 1512
- Bias = 28

### FC3
- Input : 28
- Output : 10
- Num of Parameters = 28 * 10 = 280
- Bias = 10

### Total : 42336 + 54 + 1512 + 28 + 280 + 10 = 44220